### PR TITLE
Add javascript redirect for some routes

### LINKED ISSUE #7467: Long redirect URLs cause tsh login to fail
**Problem**

Use reported seeing the following errors when attempting to use `tsh login` with their SSO provider. Web UI login works fine.

```
ERRO [WEB]       Error while processing callback. auth:oidc error:[
ERROR REPORT:
Original Error: *trace.RawTrace invalid_grant: Invalid or expired code parameter.
```

This error was coming from their identity provider because the code token has already been consumed.

Upon investigation, it was discovered that the entire SSO login flow is successful until the 302 console redirect to transfer credentials from the browser to disk. This was due to the user running a middleware component that intercepts all requests and drop requests with headers larger than 8 kb and Teleport was sending a `Location` header that was 16 kb.

The middleware would accept the request, drop it, then the 302 console redirect would be tried again and this time fail at the IdP (hence the error above) but succeed because the header was now smaller than 8 kb.

User reports that this is not the middleware they are running, but other popular Cloud services, like AWS ALB, also enforce 8 kb limits on headers.

**Proposed Solution**

This process starts with the IdP issuing a `POST /webapi/saml/acs` request which eventually becomes the 302 console redirect. Instead of performing a 302 console redirect, Teleport could return a small Javascript application that uses `window.location` to perform the redirect. We use a similar approach in Application Access.

https://github.com/gravitational/teleport/blob/master/lib/web/app/redirect.go

This would solve the problem because this request would not be caught by any middleware and instead happen local to the users workstation only.

### PR DESCRIPTION
**Pull request by atburke**
Fixes #7467.

### CODE REVIEW

**Comment by russjones** [comment]
@atburke Take a look at the following: https://github.com/gravitational/teleport/blob/master/lib/web/app/redirect.go

File: lib/web/apiserver.go
Comment (alex-kovoy) [inline lib/web/apiserver.go:2414]: To be on the safe side I suggest to add some basic checks. Something like below (just to give an idea)

```golang
	if !strings.HasPrefix(redirect, "https://"){
		return trace.Wrap("invalid redirect URL")
	}

	// to ensure a valid JS string 
	// (or do URI encoding instead)
	bytes, err := json.Marshal(struct {
		Redirect string `json:"redirect"`
	}{
		Redirect: value,
	})

	fmt.Fprintf(w, "var teleport = %v;\n", string(bytes))
```

**Comment by russjones** [comment]
@atburke Can you test an older `tsh` client with a new proxy? Also did you test with `tsh`?

**Comment by atburke** [comment]
> @atburke Can you test an older `tsh` client with a new proxy? Also did you test with `tsh`?

I've tested it with `tsh` with the current version, 6.2.13, and 5.2.4, and they all work fine.

**Comment by russjones** [comment]
@alex-kovoy Can you re-review?

File: lib/web/apiserver.go
Comment (alex-kovoy) [inline lib/web/apiserver.go:2396]: lets verify that this redirect starts with a valid scheme `https` to handle the following use-case

```golang
u, err := url.Parse("javascript:alert(1)")
```

which returns true and will result in XSS on the client side

Comment (atburke) [reply lib/web/apiserver.go:2396]: At least one of the redirects in the SSO workflow uses regular http (the one that goes to the local redirect server). Is it okay then to also allow http (or maybe check the domain as well when it is http)?

**Review by alex-kovoy — APPROVED** [review]

**Comment by russjones** [comment]
@atburke Can you also create a backport to `branch/v7` PR? I'll do some testing for this tonight as well.

**Review by russjones — APPROVED** [review]

**Comment by atburke** [comment]
> How does this redirect method compare with using a meta tag? Is one safer against injection than the other?
> 
> ```
> <meta http-equiv="refresh" content="0;URL='https://example.com'" />   
> ```

As far as I'm aware, they function identically, and I don't believe that one is safer than the other.

**Comment by zmb3** [comment]
I would think with the meta approach at least you're not executing inside a `<script>` so there's less concern about arbitrary JS code being executed, but I defer to @alex-kovoy.

**Comment by atburke** [comment]
@alex-kovoy Would it be preferable to take the `<meta>` approach rather than javascript?

**Comment by alex-kovoy** [comment]
From security perspective this is the same thing as both methods are about embedding content into html so if used unwisely you can run arbitrary code in both cases but I like the `meter` approach better because it's cleaner and also is _faster_ to complete this redirect.

**Comment by alex-kovoy** [comment]
Just make sure that there is no `caching` involved (which is handled by this headers httplib.SetNoCacheHeaders(w.Header()) but I am not sure if this impacts the `meta` behavior, perhaps worth adding this 

```
<meta http-equiv="cache-control" content="no-cache"/>
```

File: lib/web/apiserver.go
Comment (zmb3) [inline lib/web/apiserver.go:2503]: Instead of using `Fprintf`, we should use package `html/template` here, which is safe against code injection.

File: lib/web/apiserver.go
Comment (zmb3) [inline lib/web/apiserver.go:2514]: We're nearly there. 

Just need to move the template parsing out (make it a package level var). 

The way we have it here is inefficient as we would be re-parsing the same template on every redirect.

File: lib/web/apiserver.go
Comment (alex-kovoy) [reply lib/web/apiserver.go:2503]: you can also remove `nonce, err := utils.CryptoRandomHex(auth.TokenLenBytes)` as this is unnecessary after the script tag has been removed.

File: lib/web/apiserver_test.go
Comment (zmb3) [inline lib/web/apiserver_test.go:490]: I would move `t.Parallel()` out of the subtests, and then you have no need for the `tt := tt` line.

**Review by zmb3 — APPROVED** [review]

**Comment by zmb3** [comment]
FYI, it looks like the backport to v7 was merged before the merge to master was approved.

When this is approved, you probably want to backport to v8 as well as v7.

**Review by nklaassen — APPROVED** [review]

**Comment by atburke** [comment]
@fspmarshall Can you review this when you have a chance?

**Review by fspmarshall — APPROVED** [review]

**Review by Moonshine1111 — APPROVED** [review]

**Pull request merged**
